Allow scaling of window

markjfisher opened the pull request:
I hadn't come across this project until one of the FujiNet guys (well the founder Thom) was showing it on his spanky pants new machine in our discord and was using desktop zoom to effect a scale of the window. So I thought I'd have a quick hack on it to see if I could get it to naturally scale with the window size, and he suggested I PR the change.

It was a lot of fun, and this is that hack!

Hopefully it's of some use. I read that you eventually want to update it to a more modern framework, but in the interim, this allows the window to be scaled naturally.

markjfisher committed:
Allow scaling of window

mgarlanger commented:
Hey Mark, I pulled down the changes and tested locally. It appears there is a major performance hit with it.  Notice the difference in time to run the view command.

I'm attaching screen recordings of the original 


https://github.com/user-attachments/assets/f3eb21b8-e16d-42bf-989c-a3a51c959498



and with the resizing


https://github.com/user-attachments/assets/b4303975-1a89-47a8-abf3-408f99429bae


Any ideas where the performance hit is coming from.

markjfisher commented:
Hi Mark, thanks for reviewing the code.
I see CoPilot came up with some good suggestions too.

I've never used a H89 system before, so was only testing with a default system where I could type things like "Radix Hexadecimal" + "In/Out" with some keystrokes to test the screen was painting correctly.
Thanks for pointing out View, I'll have a play with it and see what improvements I can make.

The issue will come down to the change from directly placing bitmaps of the direct size needed, to scaling up to a much larger virtual buffer (2720 x 2160) and then scaling back down to whatever the view screen is so the fonts look good at all sizes. That will be the core of it.

Let me have a play.

markjfisher commented:
Just to quickly feed back, I ran some tests locally and I'm finding very little difference in times, if anything the new version for me is slightly quicker. I'm not arguing that there isn't a problem, just that I didn't notice a difference when I was writing this.

| Orig times | New default | New Large | New Small |
|------------|-----------|----------------|-------------|
| 2.33       | 1.96      | 2.66           | 1.44        |
| 2.05       | 1.93      | 1.71           | 2.02        |
| 2.46       | 1.73      | 1.87           | 1.80        |
| 1.74       | 1.96      | 1.80           | 1.87        |
| 1.96       | 1.79      | 1.71           | 1.68        |
| 1.64       | 1.71      | 1.67           |             |
| 1.85       | 1.62      | 2.41           |             |
| 2.07       | 1.69      | 1.73           |             |
| 1.75       | 1.68      | 1.60           |             |
| 1.72       | 1.98      | 1.42           |             |
| **1.957**  | **1.805** | **1.858**      | **1.762**   |

Do you have accelerated graphics on the machine you're testing on?
My machine is pretty beefy, so completely accept this could be happening on lower spec machines due to the new rendering technique. On accelerated machine I see no difference in times, but with software rendering this will take a hit.

I'll have a go at seeing if this can be made faster, but it may be we have a switch to enable scaling or not for the user so they can pick if it tries using OpenGL scaling or not.

markjfisher commented:
Also, do subsequent View commands show consistent slowness on the scaled version? I sometimes had outliers on the initial page 00 read, but definitely not as much as you were seeing here.

I'm just going through the copilot suggestions and looking at if there's a way to make it more performant without the switch

markjfisher committed:
Address copilot comments and improve performance

Instead of using 16x over sampling size, we drop to 4x
which is just as good for our purposes, and reduces the memory
footprint too.

Using nearest-neighbour instead of linear interpolation also increases
performance on non accelerated machines.

Add a caching technique so that repeated chars (e.g. spaces on screen)
don't recall the expensive glBindTexture, and reuse previous
calculations. This massively reduces number of calls for normal screen
usage.

markjfisher commented:
Hi Mark, I've added some changes to address the copilot comments, and hopefully increased performance too. They show a small benefit here, but in theory they should help a lot on non-accelerated machines.

Can you try them out please? I've add comments in the git commit for what the changes were to help performance, but TL;DR:

>     Address copilot comments and improve performance
> 
>     Instead of using 16x over sampling size, we drop to 4x
>     which is just as good for our purposes, and reduces the memory
>     footprint too.
> 
>     Using nearest-neighbour instead of linear interpolation also increases
>     performance on non accelerated machines.
> 
>     Add a caching technique so that repeated chars (e.g. spaces on screen)
>     don't rerun the expensive glBindTexture, and reuse previous
>     calculations. This massively reduces number of calls for normal screen
>     usage.

If this is still quite slow, I can add cli switches to either allow tweaking the values so the user can still get some benefits of scaling, or flat out use the previous techniques with no scaling.

markjfisher committed:
Preserve aspect ratio by default

Added a command line arg "-a" to "allow stretching"
which overrides the default of preserving the screen
aspect ratio.

markjfisher commented:
I couldn't help myself. I've also pushed a change to preserve the aspect ratio by default, so the screen doesn't stretch when resizing, but grows with aspect ration in tact.
Added a command line option "-a" to "allow stretching". I felt the default of keeping aspect ratio was nicer, so the switch disables it.

mgarlanger commented:
I'm on a MacBook Pro M1 Max. Overall system performance is great, but apple has deprecated openGL, so maybe it lacks some optimizations. 

Additional view commands seem to have the same performance issue.

With your changes, it looks like there is a minor improvement, but still significantly slower than the non-scaling code. 

Another change I noticed between your initial scaling code and the latest, is that the pixel artifacts are significant even prior to making any change in the window size. With your first change, the pixel accuracy looked decent at startup. Here is a screen shot at the initial window size.

<img width="1018" height="826" alt="Screenshot 2025-09-28 at 12 52 54 PM" src="https://github.com/user-attachments/assets/f318d017-0cfd-42b3-bc2f-c5e60081d7fe" />

mgarlanger commented:
Here is it without the resizing code.

<img width="680" height="530" alt="Screenshot 2025-09-28 at 2 28 37 PM" src="https://github.com/user-attachments/assets/ad885e5a-c3ac-41c6-9f55-211aa3f11b23" />

markjfisher commented:
what size is the window there?
On my linux, with the defaults I put in, it looks like this:

<img width="1035" height="823" alt="image" src="https://github.com/user-attachments/assets/9ebd5c53-8136-43c5-88ee-85834718a4b7" />

The reason the pixellation changed in the recent "performance improvement" patch was I went from linear texture filtering to nearest neighbour. That's slightly worse but faster. But if you're on a mac book pro m1 max, it shouldn't be an issue!

You can try undoing these changes to see if the visual improves:

<img width="927" height="168" alt="image" src="https://github.com/user-attachments/assets/50817d0c-4f6e-4374-8c1a-76aec3d0e443" />

(i.e. change GL_NEAREST to GL_LINEAR on the 2 lines)

markjfisher commented:
My AI is suggesting your screen issue (low pixel look) might also be a Retina/HiDPI display issue.
It's just suggesting some changes, let me get a patch ready.

markjfisher committed:
Add retina detection, filter mode, and help output

markjfisher commented:
I've add a few changes, you can specify `./v89 -h` now to get help.

I've added a `-f` filter flag to specify `n` for nearest, `l` for linear. It defaults back to linear as that is smoother.
It also should do some retina detection if you compile it on an apple device. I could only semi test this with some additional flags but it really needs a retina display to test it on (and I was making an assumption that was part of the issue here).

Does this improve your initial display?

markjfisher commented:
I've created some automation testing to run the View command from the application directly and by tracking the cursor location, it can time how long it's taking to run both.

I've also created a new branch and rebuilt the GUI entirely with ImGUI and SDL2 so it should render with acceleration on all platforms. The GLUT version is still there, and you'll be able to create versions with whatever engine you want.

I'm trying to tweak that to get it on par with the original bitmap implementation, as currently it's going from 1.4s in the current implementation to 1.8s with the texture mapping, and 2.0s with ImGUI. So still some investigations to make to get it on par.

EDIT: well, locally I've got the SDL2 rendering down to 1.3s by turning off the vysnc, so faster than the original implementation on my machine. I'm just tidying this up, as it might be better to ignore this PR and take the new SDL rendering in another, rather than trying to flog more out of GLUT when it has issues on Apple anyway.

markjfisher commented:
Sorry, I was trying to rename my branch so that I could revert master, but it didn't go well. I'm opening a new PR anyway, as I think this change has hit a road bump with the OpenGL issues in apple.
